content/blog/2022-04-28-chlorophyll/index.md

```diff
@@ -0,0 +1,309 @@
+  for (i from 5 to 6) x[i] = x[i] + x[i-1]; // x[i] at 0, x[i-1] at 1
+  for (i from 6 to 10) x[i] = x[i] + x[i-1]; // x[i] at 1, x[i-1] at 1
+  ```
+
+2. <u>Create symbolic variables for unannotated variables.</u> If the partition type can be inferenced from the user-annotated types, then the compiler will directly add that type after the operator or variable. Otherwise, it will generate a symbolic variable for those unannotated variables like `sym0` shown below.
+  ```cpp
+  // before annotating
+  int@6 r = 0;
+  for (i from 0 to 64) {
+    z[i] = leftrotate(x[i], y[i], r) - @place(z[i]) 1;
+    r = r +@6 1; // + happens at partition 6.
+    if (r > 32) r = 0; 
+  }
+  // after annotating
+  int@6 r = 0;
+  for (i from 0 to 64) {
+    z[i] = leftrotate(x[i]!sym0, y[i]!sym1, r!sym2) - @place(z[i]) 1;
+    r = r +@6 1; // + happens at partition 6.
+    if (r >@6 32) r = 0; 
+  }
+  ```
+
+3. <u>Construct communication count and partition space constraints with [Rosette](http://emina.github.io/rosette/) synthesizer.</u> The overall memory space should satisfy the following constraints:
+
+\\[ \text{Space} (\text{operation}) + \text{Space} (\text{statement}) + \text{Space} (\text{maxComm}) < \text{Memory per core} \\]
+
+The synthesizer will solve the constraint, use that solution as an additional constraint for `maxComm`, and then iteratively solve it to find the minimum communication cost. Finally, we can obtain the following fully annotated program.
+
+![](fully-annotated.png)
+
+One question here is where is the boundary between programmers and compilers. For this programming model, programmers do not exactly know where they should annotate the partition types. Some may benefit for the synthesizer to quickly solve the constraint, but others may have a negative impact or even cause an infeasible solution. This actually can be a burden for programmers since they still have to have a good understanding of the underlying architecture and annotate the partition in a right way.
+
+### Layout and Routing
+The next stage is to map the logical representation to physical cores. We can denote $F$ as the logical facilities or the code partition, $L$ as the physical core locations (represented in a 2D tuple), $t:F\times F\to\mathbb{R}$ as the flow function (i.e., the number of messages between two partitions), and $d:L\times L\to\mathbb{R}$ as the distance function (use Manhattan distance as a measurement). Thus, the layout and routing problem becomes finding the assignment that minimizes the following communication function:
+
+\\[ \sum_{f_1\in F, f_2\in F} t(f_1,f_2) \cdot d(a(f_1),a(f_2)) \\]
+
+This is a [Quadratic Assignment Problem (QAP)](https://en.wikipedia.org/wiki/Quadratic_assignment_problem) and can be efficently solved by [simulated annealing](https://en.wikipedia.org/wiki/Simulated_annealing#:~:text=Simulated%20annealing%20(SA)%20is%20a,space%20for%20an%20optimization%20problem.).
```

Comment:
The bit after the `#` in this URL looks a little weird—it didn't seem to do anything for me when I followed the link. Maybe it should be trimmed off?